tests/src/providers/testqgsvirtualrasterprovider.cpp

```diff
@@ -0,0 +1,232 @@
+    QVERIFY( ! layer->isValid() );
+    qDebug() << QStringLiteral( "The dataprovider is not valid" );
+  }
+
+}
+
+void TestQgsVirtualRasterProvider::testConstructor()
+{
+  QString str1 = QStringLiteral( "?crs=EPSG:4326&extent=18.6662979442000001,45.7767014376000034,18.7035979441999984,45.8117014376000000&width=373&height=350&formula=\"dem@1\" + 200&dem:provider=gdal" );
+  QString uri1 = QString( "%1&%2" ).arg( str1, QStringLiteral( "dem:uri=" ) % mTestDataDir % QStringLiteral( "raster/dem.tif" ) );
+  std::unique_ptr< QgsRasterLayer > layer_1 = std::make_unique< QgsRasterLayer >( uri1,
+      QStringLiteral( "layer_1" ),
+      QStringLiteral( "virtualraster" ) );
+
+  QVERIFY( layer_1->dataProvider()->isValid() );
+  QVERIFY( layer_1->isValid() );
+
+  double sampledValueCalc_1 = layer_1->dataProvider()->sample( QgsPointXY( 18.67714, 45.79202 ), 1 );
+  double sampledValue = mDemRasterLayer->dataProvider()->sample( QgsPointXY( 18.67714, 45.79202 ), 1 );
+
+  QCOMPARE( sampledValueCalc_1, sampledValue + 200. );
+  QCOMPARE( layer_1->dataProvider()->crs(), QgsCoordinateReferenceSystem( QStringLiteral( "EPSG:4326" ) ) );
+
+
+  QString str2 = QStringLiteral( "?crs=EPSG:32633&extent=781662.375,3339523.125,793062.375,3350923.125&width=200&height=200&formula=\"landsat@1\" + \"landsat@2\"&landsat:provider=gdal" );
+  QString uri2 = QString( "%1&%2" ).arg( str2, QStringLiteral( "landsat:uri=" ) % mTestDataDir % QStringLiteral( "landsat.tif" ) );
+  std::unique_ptr< QgsRasterLayer > layer_2 = std::make_unique< QgsRasterLayer >( uri2,
+      QStringLiteral( "layer_2" ),
+      QStringLiteral( "virtualraster" ) );
+
+  QVERIFY( layer_2->isValid() );
+  QVERIFY( layer_2->dataProvider()->isValid() );
+  QCOMPARE( layer_2->dataProvider()->crs(), QgsCoordinateReferenceSystem( QStringLiteral( "EPSG:32633" ) ) );
+  double sampledValueCalc_2 = layer_2->dataProvider()->sample( QgsPointXY( 790688, 3349113 ), 1 );
+  //qDebug() << layer_2->dataProvider()->sample( QgsPointXY( 790688, 3349113 ), 1 );
+  QCOMPARE( sampledValueCalc_2, 267. );
+
+  //use wrong formula
+  QString str3 = QStringLiteral( "?crs=EPSG:32633&extent=781662.375,3339523.125,793062.375,3350923.125&width=200&height=200&formula=\"landsat@1\" xxxxxx+ \"landsat@2\"&landsat:provider=gdal" );
+  QString uri3 = QString( "%1&%2" ).arg( str3, QStringLiteral( "landsat:uri=" ) % mTestDataDir % QStringLiteral( "landsat.tif" ) );
+  std::unique_ptr< QgsRasterLayer > layer_3 = std::make_unique< QgsRasterLayer >( uri3,
+      QStringLiteral( "layer_3" ),
+      QStringLiteral( "virtualraster" ) );
+  QVERIFY( ! layer_3->isValid() );
+
+}
+
+void TestQgsVirtualRasterProvider::testProviderProperties()
+{
+  QString formula( "\"landsat@1\" + \"landsat@2\"-\"landsat@3\"" );
+  QStringList rLayers = QgsRasterCalcNode::referencedLayerNames( formula );
+  QStringList raserRef = QgsRasterCalcNode::cleanRasterReferences( formula );
+  qDebug() << rLayers;
+  qDebug() << raserRef;
```

Comment:
Also here, make sure to check test outputs against expected values using QVERIFY / QCOMPARE